### ISSUE TITLE
Support for function parameter completion

### ISSUE THREAD

**Issue by aravind2612krishna** [natebosch/vim-lsc#82]
First of all, thanks for the great work! 

It would enormously bump up the usefulness of this plugin, I'm sure for a lot of people including me, if the completion included support for snippets that would help filling in the parameters. This gif might better explain my idea.
![687474703a2f2f692e696d6775722e636f6d2f6d7873713959792e676966](https://user-images.githubusercontent.com/3424615/39436694-5dca643a-4cbc-11e8-812e-c90c1db96be1.gif)

**Comment by natebosch** [comment]
Yeah this would be awesome. It's something I had in the back of my mind when I was working on autocomplete but I never got brave enough to tackle it.

LSP does have a spec for completing with snippets: https://microsoft.github.io/language-server-protocol/specification#snippet-syntax

I think we could rely on ultisnips for the template expansion and handling tabstops:
https://github.com/SirVer/ultisnips/blob/6fdc3647f72e0a1f321ea6bd092ecd01f7c187ba/doc/UltiSnips.txt#L428

The syntax is already pretty close between ultisnips and LSP, but we'd need to deal with the edges where they differ. For instance LSP supports [choice tabstops](https://microsoft.github.io/language-server-protocol/specification#choice) and UltiSnips doesn't understand that syntax.

There is a nice discussion about this here: https://github.com/SirVer/ultisnips/issues/964

**Comment by andreyorst** [comment]
May I ask a question about integration LSP snippets with snippet managers? I'm interested in general mechanics of how such language clients are handling this.

I've [developed my snippet solution](https://github.com/andreyorst/SimpleSnippets.vim), and I've made (just in case, i'm interested in integrating with LSP) a function, that will allow other plugins to add temporary (per session) snippets for further usage. So if LSP finds that he can give some snippet to a user, does it define a snippet as a pair of strings (trigger, definition), so it could be passed to snippet manager, which accepts a string definition of a snippet? The function for my plugin is `SimpleSnippets#addFlashSnippet(trigger, definition)`, so language client can call it for example and add (or remove) snippets:
```vim
:call SimpleSnippets#addFlashSnippet('fun', 'fun(${1:a}, ${2:b})')
:call SimpleSnippets#removeFlashSnippet('fun')
```
Or does LSP stores snippets somewhere so my plugin is the one who should constantly look for them somewhere in the memory?

As for [choice](https://microsoft.github.io/language-server-protocol/specification#choice), I think I can add this syntax fairly easily in a few hours.

**Comment by damnskippy** [comment]
FWIW, if this helps as a reference, the [clang_complete](https://github.com/Rip-Rip/clang_complete) plugin does something like this.  I've used it extensively for quite a while, and it works pretty smoothly.
Link to the [doc](https://github.com/Rip-Rip/clang_complete/blob/0918788ea0b9dc4c753ffd162c95f890ae57a275/doc/clang_complete.txt#L128)

**Comment by andreyorst** [comment]
As I said,  I've implemented `choice` placeholders:
![choice](https://user-images.githubusercontent.com/19470159/40886215-bd004fe2-673c-11e8-9574-547faa7b7471.gif)
Upon the jump the placeholder is cleared so user could write something else or select from the list.

**Comment by natebosch** [comment]
@andreyorst - what do you think about the approach that neosnippet took here? With #142 and an autocmd on `CompleteDone` would that be enough to hook it up for completion with SimpleSnippets?
